### ISSUE TITLE
 Collector eBPF probe evaluation with custom eBPF probe (on s390x) 

### ISSUE THREAD

**Issue by hbrueckner** [stackrox/collector#923]
## Overview
With the eBPF enablement for s390x in Falco libs (https://github.com/hbrueckner/falcosecurity-libs/tree/mauro-pull-libs-20221115-s390x-bpf), I looked further into the custom eBPF probe, introduced with #670. There I experienced few issues that I'd like to discuss here.

My approach to get familiar and make the [custom eBPF](https://github.com/stackrox/collector/blob/master/kernel-modules/probe/collector_probe.c#L11) by replacing the Falco libs probe directly with the custom probe and using `sinsp-example` to install and capture events.

### BPF_SUPPORTS_RAW_TRACEPOINTS
The `BPF_SUPPORTS_RAW_TRACEPOINTS` macro is explicitly undefined in [line 12](https://github.com/stackrox/collector/blob/50e9b185b4d2f9f9e2ddac4a968f362241028362/kernel-modules/probe/collector_probe.c#L12). While this has effects on the raw syscalls enter/exit (catch-all), this also disables the functionality of the [`sched_process_fork`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/trace/events/sched.h#n372) trace point.  This tracepoint is actually required for [s390x and aarch64](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/feature_gates.h#L63) to capture newly created child process information as both architecture do not generate in the `fork`/`vfork`/`cloneX` system calls. See [here](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/feature_gates.h#L36) for more details. Finally, my question here is whether collector and the upper layer require to obtain child process events?

Additional background: The raw tracepoints support is required to get access to the TP_PROTO arguments, in particular, to the `struct task_struct` of the parent and child processes. Those are required by the respective Falco libs [filler](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/fillers.h#L6587) ([child](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/fillers.h#L6826) and [parent](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/fillers.h#L6827)).

### COLLECTOR_PROBEs
Looking into the those [collector probes](https://github.com/stackrox/collector/blob/50e9b185b4d2f9f9e2ddac4a968f362241028362/kernel-modules/probe/collector_probe.c#L125) and reviewing related PR #670, I have few questions on that.

Instead of using the raw system call enter / exit events, those `COLLECTOR_PROBE` install BPF programs directly on the specified system call, e.g., `chdir`. This approach is very nice to focus on specific system calls rather hooking into every system call.

Trying those on s390x (including fixing few BPF verifier issues), resulted in this error message:
````
# ./libsinsp/examples/sinsp-example -b  driver/bpf/probe.o -j
-- Try to open: 'bpf' engine.
driver/bpf/probe.o
terminate called after throwing an instance of 'scap_open_exception'
  what():  PERF_EVENT_IOC_SET_BPF: Permission denied
Aborted (core dumped)
````
Looking into this issue, [attaching the program to the syscall](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/userspace/libscap/engine/bpf/scap_bpf.c#L645) fails and further looking into the [respective kernel sources](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/events/core.c#n10271), this issue seems to relate that the BPF program tries to data beyond the defined tracepoint structure (format).

For the `chdir`, the tracepoint structure format for its enter and exit is defined as follows:
````
# cat /sys/kernel/tracing/events/syscalls/sys_enter_chdir/format 
name: sys_enter_chdir
ID: 604
format:
        field:unsigned short common_type;       offset:0;       size:2; signed:0;
        field:unsigned char common_flags;       offset:2;       size:1; signed:0;
        field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
        field:int common_pid;   offset:4;       size:4; signed:1;

        field:int __syscall_nr; offset:8;       size:4; signed:1;
        field:const char * filename;    offset:16;      size:8; signed:0;

print fmt: "filename: 0x%08lx", ((unsigned long)(REC->filename))

# cat /sys/kernel/tracing/events/syscalls/sys_exit_chdir/format 
name: sys_exit_chdir
ID: 603
format:
        field:unsigned short common_type;       offset:0;       size:2; signed:0;
        field:unsigned char common_flags;       offset:2;       size:1; signed:0;
        field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
        field:int common_pid;   offset:4;       size:4; signed:1;

        field:int __syscall_nr; offset:8;       size:4; signed:1;
        field:long ret; offset:16;      size:8; signed:1;

print fmt: "0x%lx", REC->ret
````
This basically translates to those structure definition (there are similar the [raw system call enter exit structures](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/types.h#L32)):
````c
struct sys_enter_chdir_args
{
        __u64 pad;
        int __syscall_nr;
        const char *filename;

struct sys_exit_chdir_args
{
        __u64 pad;
        int __syscall_nr;
        long ret;
};
````
These structures are passed as context (`ctx`) to the `enter_probe` and `exit_probe` functions when the tracepoint is triggered. However, those tracepoint structure are syscall specific. The Falco libs implementation however deals with [`struct sys_enter_args`](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/types.h#L32) and [`struct sys_exit_args`](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/types.h#L45). Looking specifically to the [`struct sys_exit_args`](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/types.h#L45):
````c
struct sys_exit_args {
	__u64 pad;
	long id;
	long ret;
};
````
and comparing with the `struct sys_exit_chdir_args` above, accessing`ret` (return value) would read beyond the `struct sys_exit_chdir_args` structure because the Falco probe and filler implementation works on  `struct sys_exit_args`. This seems to explain the ` PERF_EVENT_IOC_SET_BPF: Permission denied` from above.

With above background on syscall-specific structures vs. Falco implementation on `struct sys_{enter,exit}_args`, I tried to map the syscall-specific structure into the syscall enter structures, e.g. for `chdir`.

In [`probe_enter`](https://github.com/stackrox/collector/blob/50e9b185b4d2f9f9e2ddac4a968f362241028362/kernel-modules/probe/collector_probe.c#L270), perform the conversion before storing/stashing the arguments:
````c
  //memcpy(stack_ctx.args, ctx->args, sizeof(ctx->args));
  switch (id)   /* maybe use sc_evt here..... */
  {
  case __NR_chdir: {
        struct sys_enter_chdir_args *chdir_args = (struct sys_enter_chdir_args *)ctx;
        stack_ctx.args[0] = (unsigned long)chdir_args->filename;
        break;
  }
  }
````
With that change, the original `ctx` is preserved and the `stack_ctx` is a `struct sys_enter_args` structure containing the arguments from the syscall specific tracepoint format. Also note that the `ctx` needs to be passed down the stack, especially to `call_filler` and `fork` kind of filler as they requires the context to perform the `bpf_tail_call` to subsequent filler implementation.

Using this approach (and limiting here for the enter event only), results in:
````
./libsinsp/examples/sinsp-example -b  driver/bpf/probe.o -j :
{"evt.args":"","evt.cpu":1,"evt.dir":">","evt.num":5,"evt.time":1669651102819446242,"evt.type":"chdir","proc.name":"bash","thread.tid":1056}
````

For the `probe_exit` function, this becomes more difficult because the system call specific tracepoint format does not match the generic system call exit args structure (`int` vs. `long`). Also Falco fillers use [`bpf_syscall_get_retval`](https://github.com/stackrox/falcosecurity-libs/blob/754bc65b9c569bef07d7a23f88c27f53b316c2a4/driver/bpf/plumbing_helpers.h#L82) that operates on `ctx` and casts it to `struct sys_exit_args`.

### Summary
Based on the above observations, and trying different variations with the custom probe, here is my summary for s390x:

- No raw tracepoint support (`BPF_SUPPORTS_RAW_TRACEPOINTS`) + `COLLECTOR_PROBE` has a system call specific context structure different from Falco assumption. (See above).

- No raw tracepoint support (`BPF_SUPPORTS_RAW_TRACEPOINTS`) + raw_syscalls/sys_{enter,exit} aka. `COLLECTOR_LEGACY_PROBE`: Works with some BPF verifier fixes but `sched_process_fork` filler does not work due to filler requirements to access the arguments of the tracepoint definition (= TP_PROTO). Without `sched_process_fork`, there will be no child process events on s390x (and aarch64). Also it has some performance effects because of generic system call probes (catch-all).

- Supporting raw tracepoints (`BPF_SUPPORTS_RAW_TRACEPOINTS`) + `COLLECTOR_LEGACY_PROBE`. Raw tracepoint support also enables `sched_process_fork tracepoint`. This is more aligned how falco libs works but, of course, has some performance effects because of generic system call probes (catch-all) -- potentially a bit less due to raw tracepont support.

I know this a quite large documentation on my observations and hope that the details help to understand it.

cc: @Molter73 @kcrane @robbycochran

**Comment by Molter73** [comment]
Thanks for the detailed issue @hbrueckner! I believe we've found a similar situation in the past when trying to implement the custom probe on RHEL 7 and we deferred to using the regular Falco probe for that system. We did this with the consideration that RHEL 7 is a relatively old system that should be phased out eventually, this changes if newer OSes are facing the same issue on other architectures like s390x and we should probably work out a proper solution for it.

I believe the `No raw tracepoint` would be our preferred approach, since it gives us a pretty significant performance gain. The thing I failed to understand is why does this fail on s390x but works on x86. Looking into my x86 VM, the format for the arguments look the same as the ones you pasted.

It'd be nice to have draft PRs with the full extent of the changes to see how much would this complicate the code. I know you are waiting on #912, I'll try to get it merged today if CI collaborates.

@Stringy would like to hear your thoughts since you worked on the custom probe implementation.

**Comment by Stringy** [comment]
I think it is best to treat the current Falco definitions of `sys_enter_args` and `sys_exit_args` as x86 versions of those structures, especially in the no raw tracepoints case. If we were to define (probably in falco itself) s390x versions of those structures (based on the content of `/sys/kernel/debug/tracing/events/raw_syscalls/sys_{enter,exit}/format`) then I would expect the permission denied error to go away.

The reason it works currently is because the exit structure is the same for all syscall-specific tracepoints, and the sys_enter_* structures are always <= the size of `sys_enter_args` and aligned nicely so Falco can pull out specific args without knowing the difference

**Comment by hbrueckner** [comment]
> I think it is best to treat the current Falco definitions of `sys_enter_args` and `sys_exit_args` as x86 versions of those structures, especially in the no raw tracepoints case. If we were to define (probably in falco itself) s390x versions of those structures (based on the content of `/sys/kernel/debug/tracing/events/raw_syscalls/sys_{enter,exit}/format`) then I would expect the permission denied error to go away.

Actually the falco defintions match the definitions of the raw_syscalls:
````
# cat /sys/kernel/tracing/events/raw_syscalls/sys_{enter,exit}/format 
name: sys_enter
ID: 289
format:
        field:unsigned short common_type;       offset:0;       size:2; signed:0;
        field:unsigned char common_flags;       offset:2;       size:1; signed:0;
        field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
        field:int common_pid;   offset:4;       size:4; signed:1;

        field:long id;  offset:8;       size:8; signed:1;
        field:unsigned long args[6];    offset:16;      size:48;        signed:0;

print fmt: "NR %ld (%lx, %lx, %lx, %lx, %lx, %lx)", REC->id, REC->args[0], REC->args[1], REC->args[2], REC->args[3], REC->args[4], REC->args[5]
name: sys_exit
ID: 288
format:
        field:unsigned short common_type;       offset:0;       size:2; signed:0;
        field:unsigned char common_flags;       offset:2;       size:1; signed:0;
        field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
        field:int common_pid;   offset:4;       size:4; signed:1;

        field:long id;  offset:8;       size:8; signed:1;
        field:long ret; offset:16;      size:8; signed:1;

print fmt: "NR %ld = %ld", REC->id, REC->ret
````
And the `id` field differs in size between raw syscalls and syscall-specifc exit format.

**Comment by hbrueckner** [comment]
Looks like that structure adjustments (`long` to `int`) to fit the syscall-specific format along with converting the arguments in the `enter_probe` might be an approach (actually the [`memcpy`](https://github.com/stackrox/collector/blob/6a3ddccb298b8126917564dc575fcfb1f12463e9/kernel-modules/probe/collector_probe.c#L270) reads beyond the syscall-specific format).

However, this does not solve the other with `sched_process_fork`.

**Comment by Stringy** [comment]
> Actually the falco defintions match the definitions of the raw_syscalls

Damn, that is unfortunate.

The `sched_process_fork` problem is tricky. It may not be possible to benefit from the performance improvements of syscall-specific tracepoints alongside the fork events, ultimately due to Falco itself. Unless we are able to make Falco support tracepoints alongside raw_tracepoints (which I imagine is a tremendous amount of work) I don't think we can get both. 

I think, given that we should obviously focus on functionality over performance, and that we have a very good handle on what this performance impact actually is, we may have to accept that we need to run catch-all sys_enter/sys_exit probes on s390x, at least for now. With our intention to move to a single BTF-supporting probe, this impact may only be temporary.

**Comment by hbrueckner** [comment]
Hi @Stringy,

> The `sched_process_fork` problem is tricky. It may not be possible to benefit from the performance improvements of syscall-specific tracepoints alongside the fork events, ultimately due to Falco itself. Unless we are able to make Falco support tracepoints alongside raw_tracepoints (which I imagine is a tremendous amount of work) I don't think we can get both.

It would not work in a single probe, my thought was to build two separate probes, the custom one and a `sched_process_fork` one and use some tooling (`ld -r`) to combine both. I will give it a try this week.

> I think, given that we should obviously focus on functionality over performance, and that we have a very good handle on what this performance impact actually is, we may have to accept that we need to run catch-all sys_enter/sys_exit probes on s390x, at least for now.

That I guess, would be the fallback if all above will not work.

> With our intention to move to a single BTF-supporting probe, this impact may only be temporary.

Could you share some more insights (or offline in slack)? Thanks.

**Comment by Molter73** [comment]
> It would not work in a single probe, my thought was to build two separate probes, the custom one and a sched_process_fork one and use some tooling (ld -r) to combine both. I will give it a try this week.

@hbrueckner, this sounds like it will be kinda tricky to automate. Granted right now we would only need to do this downstream for the s390x probes, but on the other hand... We would need to maintain a custom build automation downstream just for the s390x probes. If we can find a solution that prevents us from building 2 probes for each s390x kernel I think that woulld be best, but if there is no other way around it, guess we'll get the downstream builds a little messy.

You can also go ahead and test it if you want, just to have a backup if everything else fails, I'm not against that :)

**Comment by hbrueckner** [comment]
Hi @Molter73 @Stringy , @kcrane 
I now have a `s390x` version running with the collector custom probes (= direct attached system calls).

The required changes are on my personal repos:
-  [libs:s390x-mauro-latest-4-collector](https://github.com/hbrueckner/falcosecurity-libs/commits/s390x-mauro-latest-4-collector)
- [collector:s390x-mauro-latest-4-collector](https://github.com/hbrueckner/stackrox-collector/commits/s390x-mauro-latest-4-collector)

The changes include already the `socketcall` handling (with filtering for desired network calls).

**Comment by msugakov** [comment]
Hi @hbrueckner 

It looks like s390x support got merged and so I'm closing this issue. In case there are any items left to address, please feel free to reopen.
